File: rb101_rb109/medium2/tri_angles.rb
=begin
input: 3 integers, specified in degrees
output: returns a symbol :right, :acute, :obtuse, or :invalid depending on whether the triangle is a right, acute, obtuse, or invalid triangle.
rules:
  -To be a valid triangle, the sum of the angles must be exactly 180 degrees, 
  and all angles must be greater than 0

=end

# first attempt

def valid_triangle?(arr_of_angles)
  arr_of_angles.inject(:+) == 180 &&
  arr_of_angles.all? {|ang| ang > 0}
end

def triangle(ang1, ang2, ang3)
  arr = [ang1, ang2, ang3]
  case
  when !valid_triangle?(arr)  
    :invalid
  when arr.include?(90)
    :right
  when arr.any? {|ang| ang > 90}
    :obtuse
  else
    :acute
  end
end

# another solution
=begin
def triangle(ang1, ang2, ang3)
  arr = [ang1, ang2, ang3]
  case
  when arr.reduce(:+) != 180, arr.include?(0)  # when one of the two condition is met, then :invalid is executed
    :invalid
  when arr.include?(90)
    :right
  when arr.any? {|ang| ang > 90}
    :obtuse
  else
    :acute
  end
end
=end
puts triangle(60, 70, 50) 
puts triangle(30, 90, 60)
puts triangle(120, 50, 10)
puts triangle(0, 90, 90)
puts triangle(50, 50, 50)

File: ls215/medium2_2nd/tri_angles.js
/*
18:45 - 18: 55

input: angles in degrees (integers)
output: one of  4 strings
rules:
  valid triangle: 
    - sum of angles === 180
    - each angle > 0
  Right: One angle is a right angle (exactly 90 degrees).
Acute: All three angles are less than 90 degrees.
Obtuse: One angle is greater than 90 degrees.

data: array of angles

algo:
  - create array of 3 angles
  - check valid triangle 
    - sum array === 180 
    - every angle > 0
  - if one angle is 90 => right
  - if one is > 90 => obtuse
  - else => acute
*/

function triangle(angle1, angle2, angle3) {
  let angles = [angle1, angle2, angle3];

  if (!validTriangle(angles)) {
    return 'invalid';
  }

  if (angles.some(angle => angle === 90)){
    return 'right';
  } else if (angles.some(angle => angle > 90)) {
    return 'obtuse';
  } else {
    return 'acute';
  }
}

function validTriangle(angles) {
  return angles.reduce((sum, angle) => sum + angle) === 180 &&
  angles.every(angle => angle > 0);
}


console.log(triangle(60, 70, 50));       // "acute"
console.log(triangle(30, 90, 60));       // "right"
console.log(triangle(120, 50, 10));      // "obtuse"
console.log(triangle(0, 90, 90));        // "invalid"
console.log(triangle(50, 50, 50));       // "invalid"